File: algorithms/javascript/counting_sort.js
/**
 *     Counting Sort Algorithm
 *
 * Counting sort is a sorting algorithm that sorts the elements of an array by counting
 * the number of occurrences of each unique element in the array.
 * The count is stored in an auxiliary array and the sorting is done
 * by mapping the count as an index of the auxiliary array.
 *
 * Explanation taken from: https://programiz.com/dsa/counting-sort
 *
 * @param {Array} list List of integers eg. [3, -12, 7, 2, -9, 1, 0, 14, 9]
 */

function countingSort(list) {
  const min = Math.min(...list);
  const max = Math.max(...list);

  let count = [];
  let a = min;
  let b = 0;

  for (a; a <= max; a++) {
    count[a] = 0;
  }

  for (a = 0; a < list.length; a++) {
    count[list[a]] += 1;
  }

  for (a = min; a <= max; a++) {
    while (count[a]-- > 0) {
      list[b++] = a;
    }
  }
  return list;
}

/**
 * Additional function to create sample array to use counting sort on
 *
 * @param {Integer} min Minimum Value
 * @param {Integer} max Maximum Value
 * @param {Integer} length Length of desired array
 * @return {Array}  List of random integers between min and max
 */
function generateRandomListOfIntegers(min, max, length) {
  return Array.from({ length }, () => {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  });
}

let list = generateRandomListOfIntegers(-50, 50, 20);

console.log("Random list:");
console.log(list);

console.log("Sorted list:");
console.log(countingSort(list));

File: algorithms/cpp/countingSort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void countingSort(vector<int>& arr) {
    int max = *max_element(arr.begin(), arr.end()); // Find the maximum element in the array

    vector<int> count(max + 1, 0); // Create a count array to store the count of each element
    vector<int> output(arr.size());

    // Count the occurrences of each element in the input array
    for (int i = 0; i < arr.size(); ++i) {
        count[arr[i]]++;
    }

    // Calculate the cumulative count
    for (int i = 1; i <= max; ++i) {
        count[i] += count[i - 1];
    }

    // Build the output array
    for (int i = arr.size() - 1; i >= 0; --i) {
        output[count[arr[i]] - 1] = arr[i];
        count[arr[i]]--;
    }

    // Copy the sorted elements back to the original array
    for (int i = 0; i < arr.size(); ++i) {
        arr[i] = output[i];
    }
}

int main() {
    int n;
    cout << "Enter the number of elements: ";
    cin >> n;

    vector<int> arr(n);
    cout << "Enter " << n << " positive elements: ";
    for (int i = 0; i < n; ++i) {
        cin >> arr[i];
    }

    countingSort(arr);

    cout << "Sorted array: ";
    for (int num : arr) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}
